hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/unittests/TestMsiTokenProvider.java

```diff
@@ -0,0 +1,102 @@
+import org.apache.hadoop.fs.azurebfs.oauth2.AzureADToken;
+import org.apache.hadoop.fs.azurebfs.oauth2.MsiTokenProvider;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.io.IOException;
+import java.util.Date;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.hamcrest.core.AllOf.allOf;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.reflect.Whitebox.getInternalState;
+import static org.powermock.reflect.Whitebox.invokeMethod;
+import static org.powermock.reflect.Whitebox.setInternalState;
+
+/**
+ * Unit test for MsiTokenProvider
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(AzureADAuthenticator.class)
+public class TestMsiTokenProvider {
+
+  private static final long ONE_HOUR = 3600 * 1000;
+  private static final long TWO_HOUR = ONE_HOUR * 2;
+
+  @Test
+  public void testMsiTokenProvider() throws Exception {
+    String testToken = "TEST_TOKEN1";
+    setMockAzureADAuthenticator(testToken,
+        System.currentTimeMillis() + TWO_HOUR);
+
+    AccessTokenProvider msiTokenProvider = new MsiTokenProvider("", "", "", "");
+    long tokenFetchTime = getInternalState(msiTokenProvider, "tokenFetchTime");
+    Assert.assertEquals(-1, tokenFetchTime);
+
+    long before = System.currentTimeMillis();
+    AzureADToken token = msiTokenProvider.getToken();
+    long after = System.currentTimeMillis();
+    long newTokenFetchTime = getInternalState(msiTokenProvider,
+        "tokenFetchTime");
+    assertThat(newTokenFetchTime,
+        is(allOf(greaterThan(before), lessThanOrEqualTo(after))));
+    assertThat(token.getAccessToken(), is(equalTo(testToken)));
+
+    assertThat(invokeMethod(msiTokenProvider, "isTokenAboutToExpire"),
+        is(false));
+
+    setInternalState(msiTokenProvider, "tokenFetchTime",
+        System.currentTimeMillis() - ONE_HOUR);
+    assertThat(invokeMethod(msiTokenProvider, "isTokenAboutToExpire"),
+        is(true));
+
+    setInternalState(msiTokenProvider, "tokenFetchTime",
+        System.currentTimeMillis() + 2 - ONE_HOUR);
+    assertThat(invokeMethod(msiTokenProvider, "isTokenAboutToExpire"),
+        is(false));
+  }
+
+  private AzureADToken setMockAzureADAuthenticator(String tokenStr, long expiry)
+      throws IOException {
+    AzureADToken token = new AzureADToken();
+    token.setAccessToken(tokenStr);
+    token.setExpiry(new Date(expiry));
+    mockStatic(AzureADAuthenticator.class);
+    when(AzureADAuthenticator
+        .getTokenFromMsi(Mockito.anyString(), Mockito.anyString(),
```

Comment:
With this mock the code which parses the response and sets the expiry from expires_on will not be hit ?

Reply:
No. Planning to do that as a separate task.